subprojects/dependency-management/src/test/groovy/org/gradle/internal/component/model/JavaEcosystemAttributeMatcherTest.groovy

```diff
@@ -0,0 +1,300 @@
+
+        def apiElements = attributes(Usage.JAVA_API, LibraryElements.JAR, 8)
+        def apiElementsClasses = attributes(Usage.JAVA_API, LibraryElements.CLASSES, 8)
+        def runtimeElements = attributes(Usage.JAVA_RUNTIME, LibraryElements.JAR, 8)
+        def runtimeElementsClasses = attributes(Usage.JAVA_RUNTIME, LibraryElements.CLASSES, 8)
+
+        def apiElements9 = attributes(Usage.JAVA_API, LibraryElements.JAR, 9)
+        def apiElementsClasses9 = attributes(Usage.JAVA_API, LibraryElements.CLASSES, 9)
+        def runtimeElements9 = attributes(Usage.JAVA_RUNTIME, LibraryElements.JAR, 9)
+        def runtimeElementsClasses9 = attributes(Usage.JAVA_RUNTIME, LibraryElements.CLASSES, 9)
+
+        def apiElements11 = attributes(Usage.JAVA_API, LibraryElements.JAR, 11)
+        def apiElementsClasses11 = attributes(Usage.JAVA_API, LibraryElements.CLASSES, 11)
+        def runtimeElements11 = attributes(Usage.JAVA_RUNTIME, LibraryElements.JAR, 11)
+        def runtimeElementsClasses11 = attributes(Usage.JAVA_RUNTIME, LibraryElements.CLASSES, 11)
+
+        when:
+        def candidates = [
+                apiElements,
+                apiElementsClasses,
+                runtimeElements,
+                runtimeElementsClasses,
+                apiElements9,
+                apiElementsClasses9,
+                runtimeElements9,
+                runtimeElementsClasses9,
+                apiElements11,
+                apiElementsClasses11,
+                runtimeElements11,
+                runtimeElementsClasses11,
+        ]
+        def matches = matcher.match(selectionSchema, candidates, requested, null, explanationBuilder)
+        then:
+        matches == [ apiElementsClasses9 ]
+    }
+
+    def "resolve compileClasspath with java-library plugin targetJvm={8,9,11} requesting 9 only primary"() {
+        def requested = attributes(Usage.JAVA_API, LibraryElements.CLASSES, 9)
+
+        def apiElements = attributes(Usage.JAVA_API, LibraryElements.JAR, 8)
+        def runtimeElements = attributes(Usage.JAVA_RUNTIME, LibraryElements.JAR, 8)
+
+        def apiElements9 = attributes(Usage.JAVA_API, LibraryElements.JAR, 9)
+        def runtimeElements9 = attributes(Usage.JAVA_RUNTIME, LibraryElements.JAR, 9)
+
+        def apiElements11 = attributes(Usage.JAVA_API, LibraryElements.JAR, 11)
+        def runtimeElements11 = attributes(Usage.JAVA_RUNTIME, LibraryElements.JAR, 11)
+
+        when:
+        def candidates = [
+                apiElements,
+                runtimeElements,
+                apiElements9,
+                runtimeElements9,
+                apiElements11,
+                runtimeElements11,
+        ]
+        def matches = matcher.match(selectionSchema, candidates, requested, null, explanationBuilder)
+        then:
+        matches == [ apiElements9 ]
+    }
+
+    private static AttributeContainerInternal attributes(String usage, String libraryElements, int targetJvm) {
```

Comment:
I think it would be beneficial to extract these attribute helper methods base attributes test fixture or trait.